scipy/stats/_continuous_distns.py

```diff
@@ -7065,17 +7067,48 @@ def _pdf(self, x, a):
         return 2.*_norm_pdf(x)*_norm_cdf(a*x)
 
     def _cdf_single(self, x, *args):
-        _a, _b = self._get_support(*args)
-        if x <= 0:
-            cdf = integrate.quad(self._pdf, _a, x, args=args)[0]
+        """
+        This method implements the method discussed in [1]. As stated in the
+        paper, the cdf value is calculated by assuming the skewness is
+        non-negative. For negative values of skewness, we use the reflection
+        property: P(x;-shape) = 1 - P(-x;shape).
+
+        As recommented in the conclusion section of [1],
+        * when shape==1, we use ``norm_cdf(Q)`` which provide exact results
+            accross all Q.
+        * we use equation 14 when Q < 0, which is an approximation for the
+            lower tail of the distribution.
+        * we use equation 3 elsewhere since it is accurate for the "central
+            portion" of the distribution (10^-20 <= cdf <= 1 - 10^-20) and
+            upper tail (cdf > 1 - 10^-20).
+
+        References
+        ----------
+        [1] Amsler, C., Papadopoulos, A. & Schmidt, P. Evaluating the cdf of
+            the Skew Normal distribution. Empir Econ 60, 3171–3202 (2021).
+            https://doi.org/10.1007/s00181-020-01868-6
+        """
+        q = x
+        (shape,) = args
+        orig_shape = shape
+
+        if orig_shape < 0:
+            shape = -shape
+            q = -q
+
+        if shape == 1:
+            # exact approximation across all values of q
+            cdf = sc.ndtr(q) ** 2
+        elif q < 0:
+            z = (1 + shape ** 2) * q * q
+            cdf = math.exp(-0.5 * z) / (math.pi * shape * z)
         else:
-            t1 = integrate.quad(self._pdf, _a, 0, args=args)[0]
-            t2 = integrate.quad(self._pdf, 0, x, args=args)[0]
-            cdf = t1 + t2
-        if cdf > 1:
-            # Presumably numerical noise, e.g. 1.0000000000000002
-            cdf = 1.0
-        return cdf
+            q = np.c_[np.zeros_like(q), q]
```

Comment:
A few thoughts here:

- It seems that this is conceptually a separate variables, so rather than re-defining `q`, let's give it a new name
- Wouldn't `[0, q]` work just fine for scalar `q`, as you've assumed above by using `math` instead of `np`? `np.array([0, q], dtype=...` if you really want to match the data type.
- Not essential, but I think now this implementation could be vectorized, at least for arrays of `q`. You might still need to iterate over the shapes. (I'm not sure whether `multivariate_normal.cdf` is vectorized over covariance matrices.)

Reply:
- Will do
- I agree, using `[0, q]` is better. That line was left over from my offline implementation that expected `q` to be an array. 
- I have improved over this implementation offline and yes, it can be vectorized for scalar parameters and array `q`. The problem lies when the shape now has to be an array. Scipy does not allow vectorizing over the covariances. See this issue I opened yesterday: https://github.com/scipy/scipy/issues/15627. Suggestions are welcome

Reply:
Yeah, I figured. Well, vectorizing over just the `x` values would be nice, as it's pretty common to need a lot of `x` values for the same shapes (e.g. to make a plot) but it's up to you.

Reply:
seems like the only chance at getting a vectorized implemention of the bivariate normal cdf is by using the method described in https://arxiv.org/abs/1004.3616